components/web_discovery/browser/web_discovery_service.cc

```diff
@@ -92,6 +99,58 @@ void WebDiscoveryService::OnConfigChange() {
   credential_manager_->JoinGroups();
 }
 
-void WebDiscoveryService::OnPatternsLoaded() {}
+void WebDiscoveryService::OnPatternsLoaded() {
+  if (!content_scraper_) {
+    content_scraper_ = std::make_unique<ContentScraper>(
+        server_config_loader_.get(), &regex_util_);
+  }
+}
+
+bool WebDiscoveryService::ShouldExtractFromPage(
+    const GURL& url,
+    content::RenderFrameHost* render_frame_host) {
+  if (!content_scraper_) {
+    return false;
+  }
+  const auto* matching_url_details =
+      server_config_loader_->GetLastPatterns().GetMatchingURLPattern(url,
+                                                                     false);
+  if (!matching_url_details) {
+    return false;
+  }
+  VLOG(1) << "URL matched pattern " << matching_url_details->id << ": " << url;
+  if (IsPrivateURLLikely(regex_util_, url, matching_url_details)) {
+    return false;
+  }
+  return true;
+}
+
+void WebDiscoveryService::StartExtractingFromPage(
+    const GURL& url,
+    mojo::Remote<mojom::DocumentExtractor> document_extractor) {
+  auto remote_id =
+      document_extractor_remotes_.Add(std::move(document_extractor));
+  content_scraper_->ScrapePage(
+      url, false, document_extractor_remotes_.Get(remote_id),
+      base::BindOnce(&WebDiscoveryService::OnContentScraped,
+                     base::Unretained(this), false));
```

Comment:
at this point it's not really clear whether `Unretained` is legit, worth adding a comment about it